File: patterns/examples/design-patterns-java/src/main/java/br/com/alura/loja/discount/Discount.java
package br.com.alura.loja.discount;

import java.math.BigDecimal;

import br.com.alura.loja.budget.Budget;

public abstract class Discount {

	protected Discount next;

	public Discount(Discount next) {
		this.next = next;
	}

	public BigDecimal calculate(Budget budget) {
		if (shouldApplyDiscount(budget)) {
			return makeCalculation(budget);
		}
		
		return next.calculate(budget);
	}
	
	public abstract BigDecimal makeCalculation(Budget budget);
	public abstract boolean shouldApplyDiscount(Budget budget);

}

File: patterns/examples/budzet/src/Discount/Discount.php
<?php

namespace Budzet\Discount;

use Budzet\Budget;

abstract class Discount
{
  public function __construct(protected ?Discount $nextDiscount)
  {
    $this->nextDiscount = $nextDiscount;
  }

  abstract public function calculate(Budget $budget): float;
}